store/sqlstore/post_store.go

```diff
@@ -1202,3 +1229,58 @@ func (s SqlPostStore) GetOldest() store.StoreChannel {
 		result.Data = &post
 	})
 }
+
+func (s *SqlPostStore) determineMaxPostSize() int {
+	var maxPostSize int = model.POST_MESSAGE_MAX_RUNES_V1
+
+	if s.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+		// PostgreSQL uses TEXT as the underlying type, and we will have safely
+		// migrated from any VARCHAR(4000) with the same underlying type on
+		// upgrade.
+		maxPostSize = model.POST_MESSAGE_MAX_RUNES_V2
+
+	} else if s.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		// Unlike PostgreSQL, MySQL's VARCHAR and TEXT types do not share the same
+		// underlying storage strategy. Honour any existing limitation, and rely
+		// on system administrators to migrate the column if necessary.
+		var maxPostSizeBytes int32
+
+		if err := s.GetReplica().SelectOne(&maxPostSizeBytes, `
+				SELECT 
+					CHARACTER_MAXIMUM_LENGTH
+				FROM 
+					INFORMATION_SCHEMA.COLUMNS
+				WHERE 
+					table_schema = DATABASE()
+				AND	table_name = 'Posts'
+				AND	column_name = 'Message'
+				LIMIT 0, 1
+			`); err != nil {
+			l4g.Error(utils.T("store.sql_post.query_max_post_size.error") + err.Error())
+		} else {
+			// Assume a worst-case representation of four bytes per rune.
+			maxPostSize = int(maxPostSizeBytes) / 4
+
+			// To maintain backwards compatibility, don't yield a maximum post
+			// size smaller than the previous limit, even though it wasn't
+			// actually possible to store 4000 runes in all cases.
+			if maxPostSize < model.POST_MESSAGE_MAX_RUNES_V1 {
+				maxPostSize = model.POST_MESSAGE_MAX_RUNES_V1
+			}
+		}
+	} else {
+		l4g.Warn(utils.T("store.sql_post.query_max_post_size.unrecognized_driver"))
+	}
+
+	return maxPostSize
+}
+
+// GetMaxPostSize returns the maximum number of runes that may be stored in a post.
+func (s *SqlPostStore) GetMaxPostSize() store.StoreChannel {
+	return store.Do(func(result *store.StoreResult) {
+		s.maxPostSizeOnce.Do(func() {
+			s.maxPostSizeCached = s.determineMaxPostSize()
+		})
+		result.Data = s.maxPostSizeCached
```

Comment:
What happens if there was an error and `s.maxPostSizeCached` is `0`? Just want to make sure we're not introducing an unrecoverable error case. Might be a good idea to keep the cache invalidated until we can get the max post size without error. Something like this maybe:

```go
type maxPostSizeCache struct {
    Size int
    Err *model.AppError
    Once sync.Once
}

type SqlPostStore struct {
    ...
    maxPostSizeCache atomic.Value
    ...
}

func (s *SqlPostStore) GetMaxPostSize() store.StoreChannel {
    return store.Do(func(result *store.StoreResult) {
        cache := s.maxPostSizeCache.Load(*maxPostSizeCache)
        cache.Once.Do(func() {
            cache.Size, cache.Err = s.determineMaxPostSize()
            if cache.Err != nil {
                s.maxPostSizeCache.Store(&maxPostSizeCache{})
            }
        })
        result.Data = cache.Size
        result.Err = cache.Err
    })
}
```

That way there's still only the minimum number of queries performed, and we'll keep trying until we succeed.

Reply:
Another option might be to mirror the error handling semantics of `determineMaxPostSize` and default `maxPostSizeCache` to `model.POST_MESSAGE_MAX_RUNES_V1`. You're right that an uncaught error (such as a panic) incorrectly leaves `maxPostSizeCache` as `0`, but if it defaulted to `model.POST_MESSAGE_MAX_RUNES_V1` and logged an error, I suspect it's better than retrying a path constantly that's likely to fail again anyway.

Reply:
Oh I missed the default of `POST_MESSAGE_MAX_RUNES_V1` in `determineMaxPostSize`. This is probably okay then.

Reply:
Sounds good -- I defaulted `maxPostSizeCache` to `POST_MESSAGE_MAX_RUNES_V1` when constructing the store anyway, just in case a panic kills the once handler and it never runs again.